openflow15/openflow15.go

```diff
@@ -0,0 +1,2536 @@
+	case Type_RequestForward:
+		message = NewRequestForward()
+	case Type_BundleControl:
+		message = NewBundleCtrl(0, 0, 0)
+	case Type_BundleAddMessage:
+		message = NewBndleAdd(0, 0)
+	case Type_ControllerStatus:
+		message = NewControllerStatusHeader()
+	default:
+		return nil, errors.New("An unknown v1.5 packet type was received. Parse function will discard data.")
+	}
+	if message != nil {
+		err = message.UnmarshalBinary(b)
+	}
+	return
+}
+
+type PacketOut struct {
+	common.Header
+	BufferId   uint32
+	ActionsLen uint16
+	pad        []byte
+	Match      Match
+	Actions    []Action
+	Data       util.Message
+}
+
+func NewPacketOut() *PacketOut {
+	p := new(PacketOut)
+	p.Header = NewOfp15Header()
+	p.Header.Type = Type_PacketOut
+	p.BufferId = 0xffffffff
+	p.ActionsLen = 0
+	p.pad = make([]byte, 2)
+	p.Match = *NewMatch()
+	p.Actions = make([]Action, 0)
+	p.Data = util.NewBuffer(make([]byte, 0))
+	return p
+}
+
+func (p *PacketOut) AddAction(act Action) {
+	p.Actions = append(p.Actions, act)
+	p.ActionsLen += act.Len()
+}
+
+func (p *PacketOut) Len() (n uint16) {
+	n += p.Header.Len()
+	n += 8 /* buffer_id, actions_len, pad[2] */
+	n += p.Match.Len()
+	for _, a := range p.Actions {
+		n += a.Len()
+	}
+	if p.Data != nil {
+		n += p.Data.Len()
+	}
+	//if n < 72 { return 72 }
+	return
+}
+
+func (p *PacketOut) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(p.Len()))
+	var b []byte
+	n := 0
+
+	p.Header.Length = p.Len()
+	b, err = p.Header.MarshalBinary()
+	if err != nil {
+		return
+	}
+	copy(data[n:], b)
+	n += len(b)
+
+	binary.BigEndian.PutUint32(data[n:], p.BufferId)
+	n += 4
+	binary.BigEndian.PutUint16(data[n:], p.ActionsLen)
+	n += 2
+
+	n += 2 /* for pad */
+
+	b, err = p.Match.MarshalBinary()
+	if err != nil {
+		return
+	}
+	copy(data[n:], b)
+	n += int(p.Match.Len())
+
+	for _, a := range p.Actions {
+		b, err = a.MarshalBinary()
+		if err != nil {
+			return
+		}
+		copy(data[n:], b)
+		n += len(b)
+	}
+
+	if p.Data != nil {
+		b, err = p.Data.MarshalBinary()
+		if err != nil {
+			return
+		}
+		copy(data[n:], b)
+		n += len(b)
+	}
+	return
+}
+
+func (p *PacketOut) UnmarshalBinary(data []byte) (err error) {
+	err = p.Header.UnmarshalBinary(data)
+	if err != nil {
+		return
+	}
+	n := p.Header.Len()
+
+	p.BufferId = binary.BigEndian.Uint32(data[n:])
+	n += 4
+	p.ActionsLen = binary.BigEndian.Uint16(data[n:])
+	n += 2
+
+	n += 2 // for pad
+
+	if err = p.Match.UnmarshalBinary(data[n:]); err != nil {
+		return err
+	}
+	n += p.Match.Len()
+	a := n
+	for n < (a + p.ActionsLen) {
+		a, err := DecodeAction(data[n:])
+		if err != nil {
+			return err
+		}
+		p.Actions = append(p.Actions, a)
+		n += a.Len()
+	}
+
+	err = p.Data.UnmarshalBinary(data[n:])
+	return err
+}
+
+// ofp_packet_in 1.5 (same as 1.3)
+type PacketIn struct {
+	common.Header
+	BufferId uint32
+	TotalLen uint16
+	Reason   uint8
+	TableId  uint8
+	Cookie   uint64
+	Match    Match
+	pad      []uint8
+	Data     util.Message
```

Comment:
Why do you change the typd of Data from Ethernet to util.Message?

Reply:
As per Openflow 1.5, 7.2.3.11
"Packet-in messages must include the Packet Type Match Field in the match, unless the packet type is Ethernet."

This means that the packet included in the Packet In message could be Ethernet or other types. Does this break any existing functionality?

Reply:
Not a blocking issue. I asked it because Antrea thought the Data filed type is Ethernet with OF1.3, but it changes when switching to OF1.5. This caused a compilation issue. But if it is as designed, I could change Antrea code to be compatible with the new field type.